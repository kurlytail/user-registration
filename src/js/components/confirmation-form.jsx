import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

class RegistrationConfirmationForm extends Component {
    constructor(props) {
        super(props);
        this.state = { hash: props.hash, email: props.email };
    }

    handlePassword(event) {
        this.setState({ password: event.target.value });
    }

    handleConfirmPassword(event) {
        this.setState({ confirmPassword: event.target.value });
    }

    handleName(event) {
        this.setState({ name: event.target.value });
    }

    onSubmit(e) {
        e.preventDefault();
        this.props.actions.deleteRegistration({ ...this.state });
    }

    render() {
        let fieldErrors = _.get(this.props.state, 'error.errors');
        fieldErrors = fieldErrors ? fieldErrors : [];

        return (
            <form className="auth-form" method="POST" id="user-registration-form" encType="utf8">
                <div className="form-group">
                    <div className="form-label-group">
                        <label>email</label>
                        <input
                            type="email"
                            className="form-control"
                            id="user-registration-email"
                            readOnly
                            value={this.state.email}
                        />
                    </div>
                    <div className="form-label-group">
                        <label>Name</label>
                        <input
                            type="text"
                            className="form-control"
                            required
                            autoFocus
                            id="user-registration-name"
                            onChange={event => this.handleName(event)}
                            value={this.state.name}
                        />
                        {fieldErrors.map(
                            errorItem =>
                                errorItem.field === 'name' ? (
                                    <p className="text-red" key={errorItem.code}>
                                        {errorItem.defaultMessage}
                                    </p>
                                ) : (
                                    ''
                                )
                        )}
                    </div>
                    <div className="form-label-group">
                        <label>Password</label>
                        <input
                            type="password"
                            className="form-control"
                            required
                            id="user-registration-password"
                            onChange={event => this.handlePassword(event)}
                            value={this.state.password}
                        />
                        {fieldErrors.map(
                            errorItem =>
                                errorItem.field === 'password' ? (
                                    <p className="text-red" key={errorItem.code}>
                                        {errorItem.defaultMessage}
                                    </p>
                                ) : (
                                    ''
                                )
                        )}
                    </div>
                    <div className="form-label-group">
                        <label>Confirm Password</label>
                        <input
                            type="password"
                            className="form-control"
                            required
                            id="user-registration-confirmPassword"
                            onChange={event => this.handleConfirmPassword(event)}
                            value={this.state.confirmPassword}
                        />
                        {fieldErrors.map(
                            errorItem =>
                                errorItem.field === 'confirmPassword' ? (
                                    <p className="text-red" key={errorItem.code}>
                                        {errorItem.defaultMessage}
                                    </p>
                                ) : (
                                    ''
                                )
                        )}
                    </div>
                </div>
                <div className="form-group">
                    <button
                        className="btn btn-lg btn-primary btn-block"
                        type="submit"
                        id="signup-button"
                        onClick={e => this.onSubmit(e)}
                    >
                        Confirm Registration
                    </button>
                </div>
            </form>
        );
    }
}

RegistrationConfirmationForm.propTypes = {
    hash: PropTypes.string,
    email: PropTypes.string,
    state: PropTypes.any,
    actions: PropTypes.any
};

export default RegistrationConfirmationForm;
