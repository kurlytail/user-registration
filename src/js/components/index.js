import RegistrationForm from './registration-form';
import ConfirmationForm from './confirmation-form';

export { RegistrationForm, ConfirmationForm };
