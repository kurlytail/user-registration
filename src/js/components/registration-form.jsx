import React, { Component } from 'react';
import ReCAPTCHA from 'react-google-recaptcha';
import PropTypes from 'prop-types';
import _ from 'lodash';

class RegistrationForm extends Component {
    constructor(props) {
        super(props);
        this.state = {};
        this._reCaptchaRef = React.createRef();
    }

    handleEmail(event) {
        this.setState({ email: event.target.value });
    }

    onSubmit(e) {
        e.preventDefault();
        this.props.actions.createRegistration({ ...this.state, captchaKey: this.props.state.item.captchaKey });
    }

    render() {
        let fieldErrors = _.get(this.props.state, 'error.errors');
        fieldErrors = fieldErrors ? fieldErrors : [];

        return (
            <form className="auth-form" method="POST" id="user-registration-form" encType="utf8">
                <div className="form-group">
                    <div className="form-label-group">
                        <label>email</label>
                        <input
                            type="email"
                            className="form-control"
                            required
                            autoFocus
                            id="user-registration-email"
                            onChange={event => this.handleEmail(event)}
                        />
                        {fieldErrors.map(
                            errorItem =>
                                errorItem.field === 'email' ? (
                                    <p className="text-red" key={errorItem.code}>
                                        {errorItem.defaultMessage}
                                    </p>
                                ) : (
                                    ''
                                )
                        )}
                    </div>
                </div>
                <div className="form-group">
                    {_.get(this.props.state, 'item.captchaKey') && !_.get(this.props.state, 'isFetching') ? (
                        <ReCAPTCHA
                            id="user-registration-captcha"
                            sitekey={_.get(this.props.state, 'item.captchaKey')}
                            onChange={value => this.dataCallback(value)}
                            onExpired={() => this.dataExpiredCallback()}
                            onError={() => this.dataErrorCallback()}
                            ref={this._reCaptchaRef}
                        />
                    ) : (
                        <div />
                    )}
                    {fieldErrors.map(
                        errorItem =>
                            errorItem.field === 'reCaptchaResponse' ? (
                                <p className="text-red" key={errorItem.code}>
                                    {errorItem.defaultMessage}
                                </p>
                            ) : (
                                ''
                            )
                    )}
                </div>
                <div className="form-group">
                    <button
                        className="btn btn-lg btn-primary btn-block"
                        type="submit"
                        id="signup-button"
                        disabled={!this.state.reCaptchaResponse || !this.state.email}
                        onClick={e => this.onSubmit(e)}
                    >
                        Sign Up
                    </button>
                </div>
            </form>
        );
    }

    dataCallback(value) {
        this.setState({ reCaptchaResponse: value });
    }

    dataExpiredCallback() {
        this.setState({ reCaptchaResponse: null });
    }

    dataErrorCallback() {
        this.setState({ reCaptchaResponse: null });
    }
}

RegistrationForm.propTypes = {
    state: PropTypes.any,
    actions: PropTypes.any
};

export default RegistrationForm;
