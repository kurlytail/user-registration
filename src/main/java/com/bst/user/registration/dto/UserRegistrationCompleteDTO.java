package com.bst.user.registration.dto;

import javax.validation.constraints.Email;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

import com.bst.user.registration.constraints.ValidConfirmationHash;


@ValidConfirmationHash
public class UserRegistrationCompleteDTO {

	@Email
	@NotNull
	@NotEmpty
	private String email;
	
	@NotNull
	@NotEmpty
	private String name;

	@NotNull
	@NotEmpty
	private String password;
	
	@NotNull
	@NotEmpty
	private String confirmPassword;
	
	@NotNull
	@NotEmpty
	private String hash;

	public String getHash() {
		return hash;
	}

	public void setHash(String hash) {
		this.hash = hash;
	}

	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getPassword() {
		return password;
	}

	public void setPassword(String password) {
		this.password = password;
	}

	public String getConfirmPassword() {
		return confirmPassword;
	}

	public void setConfirmPassword(String confirmPassword) {
		this.confirmPassword = confirmPassword;
	}
}
