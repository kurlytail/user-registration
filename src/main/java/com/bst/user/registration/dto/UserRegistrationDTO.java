package com.bst.user.registration.dto;

import javax.validation.constraints.Email;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

import com.bst.utility.constraints.ValidReCaptcha;

public class UserRegistrationDTO {
     
    @Email
    @NotNull
    @NotEmpty
    private String email;
    
    @ValidReCaptcha
    private String reReCaptchaResponse;

	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public String getReCaptchaResponse() {
		return reReCaptchaResponse;
	}

	public void setReCaptchaResponse(String reReCaptchaResponse) {
		this.reReCaptchaResponse = reReCaptchaResponse;
	}

}
