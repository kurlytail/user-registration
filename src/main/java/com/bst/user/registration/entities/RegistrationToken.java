package com.bst.user.registration.entities;

import java.util.Date;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Transient;

import org.apache.commons.codec.digest.DigestUtils;

@Entity
public class RegistrationToken {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;
	private Date createdDate = new Date();
	@Column(unique = true)
	private String email;

	@Transient
	public String getHash() {
		return DigestUtils.sha512Hex(getEmail() + getCreatedDate() + getId().toString());
	}

	public RegistrationToken() {
		this.email = null;
	}

	public RegistrationToken(String email) {
		this.email = email;
	}

	public Long getId() {
		return id;
	}

	public Date getCreatedDate() {
		return createdDate;
	}

	public String getEmail() {
		return email;
	}
}
