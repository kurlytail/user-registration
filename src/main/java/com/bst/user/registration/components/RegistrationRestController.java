package com.bst.user.registration.components;

import java.util.Locale;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.env.Environment;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.bst.user.registration.dto.UserRegistrationCompleteDTO;
import com.bst.user.registration.dto.UserRegistrationDTO;
import com.bst.user.registration.entities.RegistrationToken;
import com.bst.utility.configuration.CaptchaSettings;
import com.bst.utility.services.EmailService;

@RestController
@RequestMapping("registration")
public class RegistrationRestController {

	@Autowired
	private CaptchaSettings captchaSettings;

	@Autowired
	private EmailService emailService;

	@Autowired
	private Environment environment;

	@Autowired
	private RegistrationService registrationService;

	@DeleteMapping
	public void completeRegistration(final @Validated @RequestBody UserRegistrationCompleteDTO userDTO) {
		this.registrationService.completeRegistration(userDTO);
	}
	
	@GetMapping
	public UserRegistrationDTO getMapping() {
		final UserRegistrationDTO userDTO = new UserRegistrationDTO();
		userDTO.setCaptchaKey(this.captchaSettings.getKey());
		return userDTO;
	}

	@PostMapping
	public void newUserRegistration(final @Validated @RequestBody UserRegistrationDTO userDTO) throws Exception {
		final RegistrationToken token = this.registrationService.commitToken(userDTO);
		this.emailService.sendMessage(new String[] { token.getEmail() },
				this.environment.getProperty("bst.email.template.user.registration.signupConfirm",
						"email/auth-signup-confirm"),
				this.environment.getProperty("bst.email.from", "automator@brainspeedtech.com"),
				"Continue your registration", "user", token, Locale.ENGLISH);
	}
}
