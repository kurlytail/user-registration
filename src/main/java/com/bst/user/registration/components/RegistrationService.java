package com.bst.user.registration.components;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.bst.user.authentication.components.UserService;
import com.bst.user.authentication.entities.User;
import com.bst.user.registration.dto.UserRegistrationCompleteDTO;
import com.bst.user.registration.dto.UserRegistrationDTO;
import com.bst.user.registration.entities.RegistrationToken;
import com.bst.user.registration.repositories.RegistrationTokenRepository;

@Service
public class RegistrationService {
	@PersistenceContext
	private EntityManager entityManager;

	@Autowired
	private RegistrationTokenRepository tokenRepository;

	@Autowired
	private UserService userService;

	@Transactional
	public RegistrationToken commitToken(UserRegistrationDTO accountDto) {

		RegistrationToken oldToken = tokenRepository.findByEmail(accountDto.getEmail());
		if (oldToken != null) {
			tokenRepository.delete(oldToken);
			entityManager.flush();
		}

		RegistrationToken token = new RegistrationToken(accountDto.getEmail());
		tokenRepository.save(token);
		entityManager.flush();
		entityManager.refresh(token);

		return token;
	}

	@Transactional
	public User completeRegistration(UserRegistrationCompleteDTO completeDto) {
		RegistrationToken token = tokenRepository.findByEmail(completeDto.getEmail());
		if (token == null) {
			return null;
		}

		User newUser = userService.createUser(completeDto.getEmail(), completeDto.getName(), completeDto.getPassword());
		tokenRepository.delete(token);
		return newUser;
	}
}
