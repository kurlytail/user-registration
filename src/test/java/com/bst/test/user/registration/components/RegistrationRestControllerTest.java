package com.bst.test.user.registration.components;

import java.util.Locale;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;
import org.mockito.Captor;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Bean;
import org.springframework.core.env.Environment;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.TestExecutionListeners;
import org.springframework.test.context.TestExecutionListeners.MergeMode;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import com.bst.test.user.registration.TestAppConfiguration;
import com.bst.user.registration.components.RegistrationRestController;
import com.bst.user.registration.components.RegistrationService;
import com.bst.user.registration.dto.UserRegistrationCompleteDTO;
import com.bst.user.registration.dto.UserRegistrationDTO;
import com.bst.user.registration.entities.RegistrationToken;
import com.bst.utility.configuration.CaptchaSettings;
import com.bst.utility.services.EmailService;
import com.bst.utility.services.ReCaptchaService;
import com.bst.utility.testlib.SnapshotListener;

@ExtendWith(SpringExtension.class)
@TestExecutionListeners(listeners = SnapshotListener.class, mergeMode = MergeMode.MERGE_WITH_DEFAULTS)
@DataJpaTest
@ContextConfiguration(classes = { TestAppConfiguration.class })
class RegistrationRestControllerTest {

	@TestConfiguration
	public static class RegistrationRestControllerTestConfiguration {
		@Bean
		public RegistrationRestController getRegistrationRestController() {
			return new RegistrationRestController();
		}
	}

	@MockBean
	private ReCaptchaService captchaService;

	@MockBean
	private CaptchaSettings captchaSettings;

	@Captor
	ArgumentCaptor<RegistrationToken> dto;

	@Captor
	ArgumentCaptor<String> dtoName;

	@Captor
	ArgumentCaptor<String> emailCaptor;

	@MockBean
	private EmailService emailService;

	@MockBean
	private Environment environment;

	@Captor
	ArgumentCaptor<String> from;
	@Captor
	ArgumentCaptor<Locale> locale;
	@Captor
	ArgumentCaptor<String> nameCaptor;
	@Captor
	ArgumentCaptor<String> passwordCaptor;
	@Captor
	ArgumentCaptor<String[]> recepients;
	@Captor
	ArgumentCaptor<UserRegistrationCompleteDTO> registrationCompleteDTOCaptor;
	@MockBean
	private RegistrationService registrationService;
	@Autowired
	private RegistrationRestController restController;
	@Captor
	ArgumentCaptor<String> subject;

	@Captor
	ArgumentCaptor<String> template;

	@BeforeEach
	public void setup() {
		Mockito.when(this.captchaSettings.getKey()).thenReturn("test captcha key");
		Mockito.when(this.registrationService.commitToken(ArgumentMatchers.any(UserRegistrationDTO.class)))
				.thenReturn(new RegistrationToken());
	}

	@Test
	public void testConfirmationPost() throws Exception {
		final UserRegistrationCompleteDTO completeDTO = new UserRegistrationCompleteDTO();
		completeDTO.setEmail("john@doe.com");
		completeDTO.setPassword("password");
		completeDTO.setName("John Doe");
		completeDTO.setConfirmPassword("password");
		this.restController.confirmUserRegistration(completeDTO);

		Mockito.verify(this.registrationService).completeRegistration(this.registrationCompleteDTOCaptor.capture());
		SnapshotListener.expect(this.registrationCompleteDTOCaptor.getValue()).toMatchSnapshot();
	}

	@Test
	public void testRegistrationDelete() throws Exception {
		final UserRegistrationCompleteDTO userDTO = new UserRegistrationCompleteDTO();
		this.restController.completeRegistration(userDTO);
		Mockito.verify(this.registrationService).completeRegistration(userDTO);
	}

	@Test
	public void testRegistrationGet() throws Exception {
		SnapshotListener.expect(this.restController.getMapping()).toMatchSnapshot();
	}

	@Test
	public void testRegistrationPost() throws Exception {
		final UserRegistrationDTO userDTO = new UserRegistrationDTO();
		this.restController.newUserRegistration(userDTO);

		Mockito.verify(this.registrationService).commitToken(userDTO);
		Mockito.verify(this.emailService).sendMessage(this.recepients.capture(), this.template.capture(),
				this.from.capture(), this.subject.capture(), this.dtoName.capture(), this.dto.capture(),
				this.locale.capture());

		SnapshotListener.expect(this.recepients.getValue()).toMatchSnapshot();
		SnapshotListener.expect(this.template.getValue()).toMatchSnapshot();
		SnapshotListener.expect(this.from.getValue()).toMatchSnapshot();
		SnapshotListener.expect(this.subject.getValue()).toMatchSnapshot();
		SnapshotListener.expect(this.dtoName.getValue()).toMatchSnapshot();
		SnapshotListener.expect(this.dto.getValue()).toMatchSnapshot();
		SnapshotListener.expect(this.locale.getValue()).toMatchSnapshot();
	}

}
