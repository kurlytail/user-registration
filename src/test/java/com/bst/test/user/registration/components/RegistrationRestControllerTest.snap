{
  "testRegistrationPost" : [ [ null ], { }, { }, "Continue your registration", "user", {
    "createdDate" : 1542059308267
  }, "en" ],
  "testRegistrationGet" : [ {
    "captchaKey" : "test captcha key"
  } ],
  "testConfirmationPost" : [ {
    "email" : "john@doe.com",
    "name" : "John Doe",
    "password" : "password",
    "confirmPassword" : "password"
  } ]
}